dnf5-plugins/copr_plugin/copr.cpp

```diff
@@ -0,0 +1,285 @@
+        if (cr.multilib)
+            std::cout << " [multilib]";
+        if (!cr.enabled)
+            std::cout << " (disabled)";
+        std::cout << std::endl;
+    };
+};
+
+const std::string hubspec_to_hostname(const std::string & hubspec) {
+    /// TODO: convert "fedora" to "copr.fedorainfracloud.org" ...
+    return hubspec;
+}
+
+
+const std::string guess_project_name_from_old_repofile_name(const std::string & repo_file_basename) {
+    auto output = std::regex_replace(repo_file_basename, std::regex("^_copr_"), "");
+    output = std::regex_replace(output, std::regex(".repo$"), "");
+    return std::regex_replace(output, std::regex("-"), "/", std::regex_constants::format_first_only);
+}
+
+
+void warn_old_copr_repofile(const std::string & repofile) {
+    /// This is a very old Copr repository format we want to warn
+    /// about, as it used to be defined here:
+    /// https://github.com/rpm-software-management/dnf-plugins-core/blob/9905d15bfcb2df00e4a0f9ad5f40453111f8f96b/plugins/copr.py#L182-L183
+    /// Even DNF4 copr.py was fixed for a long time, but e.g. the
+    /// fedora-workstation-repositories.rpm still provides an old formatted
+    /// repoffile.
+    auto base = basename(repofile);
+    if (fnmatch("_copr_*", base.c_str(), 0) != 0)
+        return;
+    warning(
+        _("Copr repofile '%s' has an old format, "
+          "fix by re-enabling it $ %s copr enable %s"),
+        repofile.c_str(),
+        dnf5_command,
+        guess_project_name_from_old_repofile_name(base).c_str());
+}
+
+
+std::string repo_id_to_copr_file(const std::string & repo_id) {
+    /// Convert repoID (generated by Copr Frontend) to a repo filename.  Note we
+    /// keep the group_ prefix for group ownernames.
+    /// - copr:copr.fedorainfracloud.org:group_copr:copr-pull-requests:pr:2545
+    /// + _copr:copr.fedorainfracloud.org:group_copr:copr-pull-requests:pr:2545.repo
+
+    auto fixed_repo_id = std::regex_replace(repo_id, std::regex(":ml$"), "");
+    return "_" + fixed_repo_id + ".repo";
+}
+
+
+void warn_weird_copr_repo(libdnf::repo::RepoWeakPtr dnfRepo) {
+    /// Throw a warning for every copr-repofile-related repo ID which doesn't
+    /// follow the pre-defined format.
+
+    std::string repo_id = dnfRepo->get_id();
+    if (fnmatch("coprdep:*", repo_id.c_str(), 0) == 0)
+        return;
+
+    auto repo_file_path = dnfRepo->get_repo_file_path().c_str();
+    if (fnmatch("copr:*", repo_id.c_str(), 0) != 0) {
+        warning(_("Repo id '%s' in '%s' is suspicious"), repo_id.c_str(), repo_file_path);
+        return;
+    }
+
+    // Check that the repo filename matches the repo ID
+    if (repo_id_to_copr_file(repo_id) != basename(dnfRepo->get_repo_file_path())) {
+        warning(_("Repo id '%s' doesn't belong to '%s' repofile"), repo_id.c_str(), repo_file_path);
+    }
+}
+
+static bool is_copr_repofile(const std::string & repofile) {
+    auto base = basename(repofile);
+    return fnmatch("_copr:*", base.c_str(), 0) == 0;
```

Comment:
Why are you using regex in some places and fnmatch in others? Is using fnmatch an optimization? What about:
`return filename.native().starts_with("_copr:");`